Fixes #25215 - Support for Debian on aarch64 aka arm64

pawelmoll opened the pull request:
Version 8 of Arm architecture defined 64-bit execution state called
AArch64. RedHat-based distributions (and GNU tools) use "aarch64" name
for it, but Debian went (as Linux kernel) with "arm64" string. This
introduced the same "schism" as happened with "x86_64" and "arm64"
and has to be handled by Debian-related routines.

This change adds handling of the arm64 case in media URL generation
patch, transforming the "arch" variable strings from "aarch64" to
"arm64" and reverses the operation for facts parsing.

Signed-off-by: Pawel Moll <pawel.moll@arm.com>

pawelmoll committed:
Fixes #25215 - Support for Debian on aarch64 aka arm64

Version 8 of Arm architecture defined 64-bit execution state called
AArch64. RedHat-based distributions (and GNU tools) use "aarch64" name
for it, but Debian went (as Linux kernel) with "arm64" string. This
introduced the same "schism" as happened with "x86_64" and "arm64"
and has to be handled by Debian-related routines.

This change adds handling of the arm64 case in media URL generation
patch, transforming the "arch" variable strings from "aarch64" to
"arm64" and reverses the operation for facts parsing.

Signed-off-by: Pawel Moll <pawel.moll@arm.com>

lzap commented on app/models/operatingsystems/debian.rb:
```diff
@@ -57,5 +57,6 @@ def guess_os
 
   def transform_vars(vars)
     vars[:arch] = vars[:arch].gsub('x86_64', 'amd64')
+    vars[:arch] = vars[:arch].gsub('aarch64', 'arm64')
```
Would you mind replacing `gsub` with `sub`? Here and also:

```
app/models/operatingsystems/debian.rb
59:    vars[:arch] = vars[:arch].gsub('x86_64', 'amd64')

app/models/operatingsystems/coreos.rb
42:    vars[:arch] = vars[:arch].gsub('x86_64', 'amd64')

app/models/operatingsystems/freebsd.rb
41:    vars[:arch] = vars[:arch].gsub('x86_64', 'amd64')
```

lzap reviewed (changes requested):
Thanks, makes sense. Minor comment before I trigger CI run.

pawelmoll committed:
Fixes #25215 - Use normal string substitution for arch strings

There's no need to use global substitution for variables transformation
when preparing media (boot file) URLs. Therefore replace all gsub-s
(leftovers from old approaches to the problem) with normal subs.

Requested-by: Lukas Zapletal
Signed-off-by: Pawel Moll <pawel.moll@arm.com>

pawelmoll commented:
Sure, done now.

mmoll commented:
ok to test

mmoll commented:
@pawelmoll please rebase onto current develop, that should fix the test failures

pawelmoll commented:
Rebased now.

mmoll reviewed (approved):
test failures unrelated :green_heart:

lzap reviewed (approved):
Thank you for the contribution. @tbrisker I think this one is trivial enough for 1.20.

lzap commented:
Apologies, @tbrisker made a good point on IRC, it is theoretically possible to have multiple variables. Can you please put `gsub` back please? Sorry about that! :)

tbrisker commented:
cherry-picked to 1.20-stable as 2037bbb and 4dfc776d81c8f32f42f4b7b32654d36a8f4f85b0

pawelmoll commented:
@lzap - I don't mind un-doing the change, but I see both patches have been just merged. So do you want a third one on top or a "revert" for the x86_64 lines and a fix for the aarch64? Also, I'm not an expert here, but how can a host's arch variable has multiple values?

lzap commented:
Sorry for the confusion, looks like miscommunication between me and @tbrisker, while we were discussing the change Tomer merged it already. He reverted the `gsub` change from the stable 1.20 branch just to make some room for more testing in nightly builds. Let's leave it for now, we will see if this is a problem.

tbrisker commented:
for the record - the new commit on the stable branch is 2b453260f3284881cb0e73b058c9f1ac5a29ec56, just to be on the safe side as we're right about to start RC process.

pawelmoll commented:
All fine with me, thanks!
